main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala

```diff
@@ -639,6 +674,99 @@ object EntryPoints {
         )
       )
     }
+
+    /**
+      * Wraps an entry point function with all necessary default effect handlers.
+      *
+      * Transforms a function:
+      * def f(arg1: tpe1, ...): tpe \ ef = exp
+      *
+      * Into:
+      * def f(arg1: tpe1, ...): tpe \ (((ef - handledEffect1) - handledEffect2) - ...) + IO =
+      * handler1(_ -> handler2(_ ->exp))
+      *
+      * The wrappers:
+      * - Removes the handled effect from the function's effect set
+      * - Add IO
+      * - Creates a lambda (_ -> originalBody) and passes it to each handler
+      * - Updates the function's type signature accordingly
+      *
+      * @param currentDef      The entry point function definition to wrap
+      * @param defaultHandlers Iterable of information about the default handler to apply
+      * @param root            The typed AST root
+      * @return The wrapped function definition with all necessary default effect handlers
+      */
+    private def wrapDefWithDefaultHandlers(currentDef: TypedAst.Def, defaultHandlers: Iterable[DefaultHandler], root: TypedAst.Root)(implicit flix: Flix): TypedAst.Def = {
+      // Obtain the concrete effect set of the definition that is going to be wrapped.
+      // We are expecting entry points, and all entry points should have a concrete effect set
+      val defEffects: CofiniteSet[Symbol.EffSym] = eval(currentDef.spec.eff) match {
+        case Result.Ok(s) => s
+        // This means eff is either not well-formed or it has type variables.
+        // Either way, in this case we will wrap with all default handlers to provide better error messages
+        case Result.Err(_) => CofiniteSet.universe
+      }
+      val necessaryHandlers = defaultHandlers.filter(h => defEffects.contains(h.handledSym))
+      necessaryHandlers.foldLeft(currentDef)((defn, handler) => wrapInHandler(defn, handler, root))
```

Comment:
We should say something about the order. Even if what we say is just that: We do not use any specific order.